### ISSUE TITLE
Assumption that User Id queries are case-insensitive

### ISSUE THREAD

**Issue by brockallen** [aspnet/Identity#1089]
I have a customer using Postgress for their DB with ASP.NET Identity and they're running into issues with the case sensitive queries against the user Id column. Is there anything that can be fixed/changed in the DbContext model builder that sets that column as case sensitive at the EF level in a DB provider agnostic way?

**Comment by divega** [comment]
cc @roji to answer if there is anything Npgsql specific you can add to the model to make the queries case insensitive.

I am curious though why the `Id` column could cause this problem. Are the entity types or the model customized to use a different mechanism to generate the key values from the default? In the default implementation `Guid.NewGuid().ToString()` is assigned on construction of a new instance and as far as I know we the exact values (with the same casing) should be used from there on in any query or update. Is there anything in the application that is switching those strings to uppercase?

**Comment by brockallen** [comment]
Yes, the Id is provided by a migration tool, IIRC.

**Comment by roji** [comment]
Is it possible to get more details here? Is the problem with the column *identifier* (i.e. `Id`) or with its contents? 

Regarding identifiers, in PostgreSQL, all unquoted identifiers are always folded to lowercase. The Npgsql EF Core emits SQL code that quotes all identifiers, making all matches case-sensitive - there's no way to change either PostgreSQL's behavior or Npgsql's. Did someone create the `Id` column with a different casing, no via EF Core?

Basically I'd need more info to be able to help here.

**Comment by brockallen** [comment]
The contents itself. The data type used in .NET is a GUID, but then it's converted to a string before using in the DB. This is not a PG issue, IMO, it's an ASP.NET Identity in its string mapping for guids as the PK.

**Comment by roji** [comment]
So if I understand correctly, the `Id` column was populated with uppercase GUID strings (by whom exactly?) but is being queried with lowercase? This seems like it would be a bug in ASP.NET identity which has to be fixed there (i.e. use same case when populating and querying) rather than trying to fix this at the database level...

I'm guessing there's no issue with SqlServer because a case-insensitive collation is set up at the database level? Unfortunately PostgreSQL doesn't support this. The only option for case-insensitive collation is by defining the column type as `citext` instead of `text` (see [the citext extension](https://www.postgresql.org/docs/current/static/citext.html)). If you can customize the model, than doing `.HasColumnType("citext")` may provide a workaround (you'd also need to do `.HasPostgresExtension("citext")`).

But again, this seems like a problem that should be fixed in ASP.NET Identity itself...

**Comment by brockallen** [comment]
ASP.NET Identity is being used to provide a new identity management system for an old, existing one. Data is being imported, so the Id is being set explicitly as part of a data migration/import from the old system. Then yes, during normal processing, queries are being done and somewhere, somehow the GUID Id being used is cased differently when a query is being done. Possibly due to ToString() on the GUID object emitting casing that differs form how their old system had it, and how it was imported into ASP.NET Identity. I can ping the customer to make any additional comments here.

**Comment by roji** [comment]
OK, thanks for the added info.

To recapitulate, the correct way to fix this is to ensure the same case is used everywhere - if I understand correctly this is more of a client issue. If a workaround is absolutely necessary at the PostgreSQL/Npgsql/EF Core level instead, it *may* be possible to redefine the column holding the data as `citext` instead of `text`.

**Comment by brockallen** [comment]
Yes, IMO, this was never a PG issue except for the fact that it's case sensitive whereas SqlServer is not (by default).

**Comment by ajcvickers** [comment]
Sorry for being slow responding--we had some internal team changes. When value conversions are implemented in EF Core 2.1, then it should be possible to put a value converter on the property that will ensure all IDs are converted to upper or lower case before being sent to the database. Until then, I think the best fix would be to run a pass on the client data and put it in a consistent state.

Closing this discussion for now, but feel free to re-open if needed.
